### PR TITLE
ddl: fix NO_ZERO_DATE handled incorrectly

### LINKED ISSUE #8637: NO_ZERO_DATE sql_mode is not handled correctly
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?

`run the following DDL both on TIBD and MySQL;`

```
CREATE TABLE a (
  id int(11) DEFAULT NULL,
  agent_start_time timestamp NOT NULL DEFAULT '0000-00-00 00:00:00'
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
```

sql_mode:
`TIDB: STRICT_TRANS_TABLES,NO_ENGINE_SUBSTITUTION
MySQL: STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,NO_ENGINE_SUBSTITUTION`

2. What did you expect to see?

TIDB: run OK.
MySQL: run failed.

3. What did you see instead?

Failure in TIDB.

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

```
Release Version: v2.1.0
Git Commit Hash: 08f0168a6caea0280d6157e5be69f2dc6fd0d5b3
Git Branch: release-2.1
UTC Build Time: 2018-11-30 11:39:09
GoVersion: go version go1.11.2 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by imtbkcat**
### What problem does this PR solve? 
fix #8637 

TiDB can't handle `NO_ZERO_DATE` mode correctly in `CREATE TABLE` statement with default value options.

In TiDB:
```
mysql> set session sql_mode='STRICT_TRANS_TABLES,NO_ZERO_DATE,NO_ENGINE_SUBSTITUTION';
Query OK, 0 rows affected (0.00 sec)

mysql> CREATE TABLE a (
    ->   id int(11) DEFAULT NULL,
    ->   agent_start_time timestamp NOT NULL DEFAULT '0000-00-00 00:00:00'
    -> ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
Query OK, 0 rows affected (0.02 sec)
```

### What is changed and how it works?

Add additional check in `checkColumnDefaultValue` function. If column type is time type and has default value, it will check SQL mode and default value.

after fix:
```
mysql> set session sql_mode='STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,NO_ENGINE_SUBSTITUTION';                Query OK, 0 rows affected, 2 warnings (0.00 sec)

mysql> CREATE TABLE a (   id int(11) DEFAULT NULL,   agent_start_time date NOT NULL DEFAULT '0000-00-00' ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
ERROR 1067 (42000): Invalid default value for 'agent_start_time'
```

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8765)

### CODE REVIEW

**Comment by imtbkcat** [comment]
PTAL @zimulala @crazycs520

File: ddl/db_integration_test.go [hunk 1]
```diff
@@ -111,6 +111,25 @@ func (s *testIntegrationSuite) TestInvalidDefault(c *C) {
 	c.Assert(terror.ErrorEqual(err, types.ErrInvalidDefault), IsTrue, Commentf("err %v", err))
 }
 
+func (s *testIntegrationSuite) TestNoZeroDateMode(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+
+	tk.MustExec("use test;")
+	tk.MustExec("set session sql_mode='STRICT_TRANS_TABLES,NO_ZERO_DATE,NO_ENGINE_SUBSTITUTION';")
```
[the remark this answers is not in the document]
Comment (imtbkcat) [reply ddl/db_integration_test.go]: ok

[the remark this answers is not in the document]
Comment (imtbkcat) [reply ddl/db_integration_test.go]: @zimulala addressed.

**Comment by imtbkcat** [comment]
@zimulala PTAL

**Comment by imtbkcat** [comment]
@zimulala @crazycs520  PTAL

File: ddl/db_integration_test.go [hunk 2]
```diff
@@ -112,6 +112,19 @@ func (s *testIntegrationSuite) TestInvalidDefault(c *C) {
...
+
+	s.testErrorCode(c, tk, "create table test_zero_date(agent_start_time date NOT NULL DEFAULT '0000-00-00')", mysql.ErrInvalidDefault)
+	s.testErrorCode(c, tk, "create table test_zero_date(agent_start_time datetime NOT NULL DEFAULT '0000-00-00 00:00:00')", mysql.ErrInvalidDefault)
+	s.testErrorCode(c, tk, "create table test_zero_date(agent_start_time timestamp NOT NULL DEFAULT '0000-00-00 00:00:00')", mysql.ErrInvalidDefault)
+
+	tk.MustExec("set session sql_mode='ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION';")
```
Comment (crazycs520) [inline ddl/db_integration_test.go]: use defer in line #120 maybe better. If the test failed, defer will also recover the `sql_mode` and won't affect another test.

Comment (imtbkcat) [reply ddl/db_integration_test.go]: @crazycs520 addressed.

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -273,6 +274,14 @@ func checkColumnDefaultValue(ctx sessionctx.Context, col *table.Column, value in
 		// In strict SQL mode or default value is not an empty string.
 		return hasDefaultValue, value, errBlobCantHaveDefault.GenWithStackByArgs(col.Name.O)
 	}
+	if value != nil && ctx.GetSessionVars().SQLMode.HasNoZeroDateMode() && ctx.GetSessionVars().SQLMode.HasStrictMode() && (col.Tp == mysql.TypeTimestamp ||
+		col.Tp == mysql.TypeDatetime || col.Tp == mysql.TypeDate) {
+		if vv, ok := value.(string); ok {
+			if vv == types.ZeroDatetimeStr || vv == types.ZeroDateStr {
```
Comment (crazycs520) [inline ddl/ddl_api.go]: What if below sql:
```sql
create table t1 (a timestamp default '0000-00-00 00');
```

Comment (imtbkcat) [reply ddl/ddl_api.go]: This sql is not allow by mysql, but pass this pr. i will fix it.

Comment (crazycs520) [reply ddl/ddl_api.go]: Please add unit test for this.

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -273,6 +274,23 @@ func checkColumnDefaultValue(ctx sessionctx.Context, col *table.Column, value in
...
+			t, err := types.ParseTime(nil, vv, col.Tp, 6)
+			if err != nil {
+				// Ignores ParseTime error, because ParseTime error has been dealt in getDefaultValue
+				// Some builtin function like CURRENT_TIMESTAMP() will cause ParseTime error.
+				return hasDefaultValue, value, nil
+			}
+			res := t.Time.Year() + t.Time.Month() + t.Time.Day()
+			res += t.Time.Hour() + t.Time.Minute() + t.Time.Second()
+			res += t.Time.Microsecond()
+			if res == 0 {
```
Comment (crazycs520) [inline ddl/ddl_api.go]: How about this:
```sql
if t.Time == types.ZeroTime {
	return hasDefaultValue, value, ErrInvalidDefaultValue.GenWithStackByArgs(col.Name.O)
}
```

File: ddl/ddl_api.go (hunk 3 above)
Comment (imtbkcat) [reply ddl/ddl_api.go]: @csuzhangxc addressed.

File: ddl/ddl.go [hunk 5]
```diff
@@ -92,6 +92,8 @@ var (
 		"unsupported drop integer primary key")
 	errUnsupportedCharset = terror.ClassDDL.New(codeUnsupportedCharset, "unsupported charset %s collate %s")
 
+	errInvalidDefaultValue = terror.ClassDDL.New(codeInvalidDefaultValue, "Invalid default value for '%s'")
```
[the remark this answers is not in the document]
Comment (imtbkcat) [reply ddl/ddl.go]: @jackysp addressed.

File: ddl/ddl_api.go [hunk 6]
```diff
@@ -274,6 +275,20 @@ func checkColumnDefaultValue(ctx sessionctx.Context, col *table.Column, value in
 		// In strict SQL mode or default value is not an empty string.
 		return hasDefaultValue, value, errBlobCantHaveDefault.GenWithStackByArgs(col.Name.O)
 	}
+	if value != nil && ctx.GetSessionVars().SQLMode.HasNoZeroDateMode() && ctx.GetSessionVars().SQLMode.HasStrictMode() && (col.Tp == mysql.TypeTimestamp ||
+		col.Tp == mysql.TypeDatetime || col.Tp == mysql.TypeDate) {
```
Comment (jackysp) [inline ddl/ddl_api.go]: It's better to use IsTypeTime

**Comment by crazycs520** [comment]
@imtbkcat Please cherry-pick to release 2.0 and release 2.1.

**Pull request merged**